### ISSUE TITLE
Support for Websockets with HTTP/2 to the backend

### ISSUE THREAD

**Issue by Kishi85** [haproxy/haproxy#162]
## What are you trying to do?
Trying to get a WebSocket (Collabora CODE in this case) to work over haproxy. This currently is officially only supported and works for h1-to-h1 WebSockets and it seems to work for h2-to-h1 WebSockets as well (although not officially supported). h2-to-h2 WebSockets fail to bootstrap with 400 Bad Request.

## What should haproxy do differently? Which functionality do you think we should add?
As mentioned in https://github.com/haproxy/haproxy/issues/131#issuecomment-509359956 haproxy currently does not support WebSockets to h2 backend servers due to support of 101 (Switching Protocols) informational status code was removed (see RFC-7540 8.1.1).

haproxy should therefore implement the extension to HTTP/2 to bootstrap WebSockets (RFC8441) to achieve feature parity between h1 and h2 backends and allow the usage of h2-to-h2 WebSockets

**Comment by lukastribus** [comment]
Which browsers support this?

**Comment by chipitsine** [comment]
Firefox 65

**Comment by Kishi85** [comment]
From what I could gather quickly without doing too much research:

Firefox (65) supports it as of late January 2019 (https://bugzilla.mozilla.org/show_bug.cgi?id=1434137)
Chrome support is/was behind a flag (https://www.chromestatus.com/feature/6251293127475200 and https://bugs.chromium.org/p/chromium/issues/detail?id=801564)

**Comment by capflam** [comment]
@Kishi85 , h1-to-h1 WebSockets works, but for h2-to-h1, I'm really surprised. h1 uses a protocol upgrade while h2 uses the CONNECT method. And there is an handshake in h1 and not in h2. I don't see how it should work. Or it is handled by your server.

**Comment by Kishi85** [comment]
It cloud be handled by Collabora CODE (only supports http/1.1) which does the 101 upgrade when NGINX (running http/2 and http/1.1) does the forward proxy (Collabora runs on a separate port and for those upgrade headers are specified in nginx) on my Nextcloud host.

So yeah maybe a bit unusual but it works. The nginx setup for reference is like this one: https://help.nextcloud.com/t/correct-nginx-proxy-pass-config/10883/2

**Comment by lukastribus** [comment]
In other words, we are talking about a feature that is supported in [less than 10% of the desktop browsers](https://en.wikipedia.org/wiki/Usage_share_of_web_browsers#Summary_tables).

I guess this feature will be useful in a few years from now, but I don't see how this would be useful today. It is a chicken or egg problem of course, but it's not like this will solve any problems at the moment.

**Comment by capflam** [comment]
Well, it is just a feature request. To be clear, it is not scheduled for the 2.1. And for now, there is no plan for future versions either. But it is still good to have a note on it.

**Comment by git001** [comment]
@Kishi85 As I understand the nginx conf poper it forces the h1 protocol.
The nginy WS solution is described in this blog post https://www.nginx.com/blog/websocket-nginx/

```
        # WebSocket support
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
```

just a wild guess but, how about to make the same with haproxy?
Example below is untested, just to get an idea.

```
(listen|backend) websocket

  mode http
  option http-use-htx

  http-request set-header Upgrade upgrade
  http-request set-header Connection upgrade # this could be tricky as the connection header is handled also by haproxy

  server ws 127.0.0.1:8080 proto http/1.1
```

**Comment by Kishi85** [comment]
@git001 that's probably exactly why it is working with http/1.1 to the backend and not h2.

**Comment by git001** [comment]
@Kishi85 what does your last comment mean? Works Websockets with the above config with haproxy or not?

**Comment by Kishi85** [comment]
@git001 It works with the configuration I provided for haproxy and nginx without your example when using http/2 or http/1.1 on the frontend and only using http/1.1 to the backend because of the connection upgrade. Using http/2 from haproxy to the backend servers breaks things because the protocol switch is unsupported in http/2.

**Comment by krm1312** [comment]
It would be nice when the backend (and frontend) both support h2 and h1 haproxy would automatically choose to use h1 when it sees Upgrade: websocket.  Right now we're looking for a way to force the backend to use h1 when we match an Upgrade header on the request.  

We can likely match a request header and choose a different backend that only declares h1 support, but, then we have to duplicate backends for each protocol version.

**Comment by wtarreau** [comment]
That's an interesting point. We currently have no way to force an HTTP version when connecting to a server. The only case where there is a choice is when using ALPN, so this means we'd need to be able to disable ALPN negotiation or better, force ALPN to a specific string. But doing the latter could break things further when mixing multiple sets of servers. So maybe we should instead consider "downgrading" the HTTP version and avoiding to send ALPN on downgraded streams. But this would also affect idle connections, as right now there's a single pool for all connections, and now this would require to classify them by version. But it can go further. Once we implement RFC8441, we might also want to classify them depending on what's supported on the server side, so it's more of a matter of connection capabilities, thus it can quickly be a multi-dimensional sort.

I think we should think about a simple initial solution that could be backported to 2.2, then a more complete solution later.

**Comment by fraenki** [comment]
I just want to ask if this is still on the roadmap to be implemented/backported to 2.2, or if it was shifted to a later release (2.4 or later?).

**Comment by a-denoyelle** [comment]
There is currently no plan to backport it as it represents quite a significant change. However, it is already implement on the 2.4 tree. If you want you can test the latest dev version to check it.

**Comment by sgohl** [comment]
I'd like to join the issue for another view on this. For me, it's not really a thing of browser support or anything, but rather of a clean haproxy config (duplicate code). In other words, I don't care particularly that HAProxy speaks h2 to the backend regarding websockets, but it should be able to switch to http/1.1 if needed

Currently, I need two separate backends just to make websockets work, but they only differ in the actual backend `proto` used

```
# haproxy version is 2.4.8-d1f8d41

frontend https
  bind :443 ssl alpn h2,http/1.1
  mode http

  option http-server-close 
  option http-use-htx

  acl hdr_connection_upgrade hdr(Connection)  -i upgrade
  acl hdr_upgrade_websocket  hdr(Upgrade)     -i websocket
  use_backend websocket if hdr_connection_upgrade hdr_upgrade_websocket

  default_backend traefik

backend traefik
  mode http
  server traefik 127.0.0.1:81 proto h2

backend websocket
  mode http
  server traefik 127.0.0.1:81
```

From a user's perspective, I would expect that HAProxy is able to just downgrade the backend connection from h2 to http/1.1 if needed, or at least have the possibility to do something like this:
```
  use_backend_proto http/1.1 if hdr_connection_upgrade hdr_upgrade_websocket
```

by the way, `proto http/1.1` does not work anymore, as it results in the error `unknown MUX protocol 'http/1.1'` hence the proto of backend websocket is left blank

**Comment by a-denoyelle** [comment]
We have just implemented a new server keyword for this named 'ws'. This allows to configure an alternative protocol for websocket tunnel. By default, its value is 'auto', which allows automatic downgrade to http/1.1 if the server ALPN contains http/1.1.

This new keyword is present in the latest development release and will be available on the upcoming 2.5. A backport to 2.4 is also scheduled.

**Comment by git001** [comment]
This feature is implemented in [[ANNOUNCE] haproxy-2.4.0](https://www.mail-archive.com/search?l=haproxy@formilux.org&q=subject:%22%5C%5BANNOUNCE%5C%5D+haproxy%5C-2.4.0%22&o=newest) .
I think this issue can be closed.

**Comment by capflam** [comment]
indeed. Thanks !
